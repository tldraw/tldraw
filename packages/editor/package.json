{
	"name": "@tldraw/editor",
	"description": "A tiny little drawing app (editor).",
	"version": "3.6.0",
	"author": {
		"name": "tldraw Inc.",
		"email": "hello@tldraw.com"
	},
	"homepage": "https://tldraw.dev",
	"license": "SEE LICENSE IN LICENSE.md",
	"repository": {
		"type": "git",
		"url": "https://github.com/tldraw/tldraw"
	},
	"bugs": {
		"url": "https://github.com/tldraw/tldraw/issues"
	},
	"keywords": [
		"tldraw",
		"drawing",
		"app",
		"development",
		"whiteboard",
		"canvas",
		"infinite"
	],
	"/* NOTE */": "These `main` and `types` fields are rewritten by the build script. They are not the actual values we publish",
	"main": "./src/index.ts",
	"types": "./.tsbuild/index.d.ts",
	"style": "./editor.css",
	"/* GOTCHA */": "files will include ./dist and index.d.ts by default, add any others you want to include in here",
	"files": [
		"editor.css"
	],
	"scripts": {
		"test-ci": "lazy inherit",
		"test": "yarn run -T jest",
		"benchmark": "yarn run -T tsx ./internal/scripts/benchmark.ts",
		"test-coverage": "lazy inherit",
		"build": "yarn run -T tsx ../../internal/scripts/build-package.ts",
		"build-api": "yarn run -T tsx ../../internal/scripts/build-api.ts",
		"prepack": "yarn run -T tsx ../../internal/scripts/prepack.ts",
		"postpack": "../../internal/scripts/postpack.sh",
		"pack-tarball": "yarn pack",
		"lint": "yarn run -T tsx ../../internal/scripts/lint.ts"
	},
	"dependencies": {
		"@tldraw/state": "workspace:*",
		"@tldraw/state-react": "workspace:*",
		"@tldraw/store": "workspace:*",
		"@tldraw/tlschema": "workspace:*",
		"@tldraw/utils": "workspace:*",
		"@tldraw/validate": "workspace:*",
		"@types/core-js": "^2.5.5",
		"@use-gesture/react": "^10.2.27",
		"classnames": "^2.3.2",
		"core-js": "^3.31.1",
		"eventemitter3": "^4.0.7",
		"idb": "^7.1.1",
		"is-plain-object": "^5.0.0",
		"lodash.isequal": "^4.5.0"
	},
	"peerDependencies": {
		"react": "^18.2.0",
		"react-dom": "^18.2.0"
	},
	"devDependencies": {
		"@peculiar/webcrypto": "^1.4.0",
		"@testing-library/jest-dom": "^5.16.5",
		"@testing-library/react": "^15.0.6",
		"@types/benchmark": "^2.1.2",
		"@types/wicg-file-system-access": "^2020.9.5",
		"benchmark": "^2.1.4",
		"fake-indexeddb": "^4.0.0",
		"jest-canvas-mock": "^2.5.2",
		"jest-environment-jsdom": "^29.4.3",
		"lazyrepo": "0.0.0-alpha.27",
		"resize-observer-polyfill": "^1.5.1"
	},
	"jest": {
		"preset": "../../internal/config/jest/node/jest-preset.js",
		"testEnvironment": "../../../packages/utils/patchedJestJsDom.js",
		"fakeTimers": {
			"enableGlobally": true
		},
		"testPathIgnorePatterns": [
			"^.+\\.*.css$"
		],
		"moduleNameMapper": {
			"^~(.*)": "<rootDir>/src/$1",
			"\\.(css|less|scss|sass)$": "identity-obj-proxy"
		},
		"setupFiles": [
			"raf/polyfill",
			"jest-canvas-mock",
			"<rootDir>/setupTests.js"
		],
		"setupFilesAfterEnv": [
			"../../internal/config/setupJest.ts"
		]
	}
}
