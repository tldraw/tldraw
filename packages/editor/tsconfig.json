{
	"extends": "../../internal/config/tsconfig.base.json",
	"exclude": ["node_modules", "dist", "**/*.css", ".tsbuild*"],
	"compilerOptions": {
		"outDir": "./.tsbuild",
		"rootDir": "src",
		"noImplicitOverride": false,
		"noImplicitReturns": false,
		"types": ["node", "@types/jest"]
	},
	"references": [
		{ "path": "../state" },
		{ "path": "../state-react" },
		{ "path": "../store" },
		{ "path": "../tlschema" },
		{ "path": "../utils" },
		{ "path": "../validate" }
	]
}
