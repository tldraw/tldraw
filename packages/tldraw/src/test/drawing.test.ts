import { TLDrawShape, TLHighlightShape, last } from '@tldraw/editor'
import { vi } from 'vitest'
import { TestEditor } from './TestEditor'
import { TEST_DRAW_SHAPE_SCREEN_POINTS } from './drawing.data'

vi.useFakeTimers()

let editor: TestEditor

afterEach(() => {
	editor?.dispose()
})

beforeEach(() => {
	editor = new TestEditor()

	editor.createShapes([])
})

type DrawableShape = TLDrawShape | TLHighlightShape

for (const toolType of ['draw', 'highlight'] as const) {
	describe(`When ${toolType}ing...`, () => {
		it('Creates a dot', () => {
			editor
				.setCurrentTool(toolType)
				.pointerDown(60, 60)
				.expectToBeIn(`${toolType}.drawing`)
				.pointerUp()
				.expectToBeIn(`${toolType}.idle`)

			expect(editor.getCurrentPageShapes()).toHaveLength(1)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.type).toBe(toolType)
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('free')
		})

		it('Creates a dot when shift is held down', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(60, 60)
				.expectToBeIn(`${toolType}.drawing`)
				.pointerUp()
				.expectToBeIn(`${toolType}.idle`)

			expect(editor.getCurrentPageShapes()).toHaveLength(1)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.type).toBe(toolType)
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('straight')
		})

		it('Creates a free draw line when shift is not held', () => {
			editor.setCurrentTool(toolType).pointerDown(10, 10).pointerMove(20, 20)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('free')
		})

		it('Creates a straight line when shift is held', () => {
			editor.setCurrentTool(toolType).keyDown('Shift').pointerDown(10, 10).pointerMove(20, 20)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(1)

			const segment = shape.props.segments[0]
			expect(segment.type).toBe('straight')

			const points = segment.points
			expect(points.length).toBe(2)
		})

		it('Switches between segment types when shift is pressed / released  (starting with shift up)', () => {
			editor
				.setCurrentTool(toolType)
				.pointerDown(10, 10)
				.pointerMove(20, 20)
				.keyDown('Shift')
				.pointerMove(30, 30)
				.keyUp('Shift')
				.pointerMove(40, 40)
				.pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(3)

			expect(shape.props.segments[0].type).toBe('free')
			expect(shape.props.segments[1].type).toBe('straight')
			expect(shape.props.segments[2].type).toBe('free')
		})

		it('Switches between segment types when shift is pressed / released (starting with shift down)', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(10, 10)
				.pointerMove(20, 20)
				.keyUp('Shift')
				.pointerMove(30, 30)
				.keyDown('Shift')
				.pointerMove(40, 40)
				.pointerUp()

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(3)

			expect(shape.props.segments[0].type).toBe('straight')
			expect(shape.props.segments[1].type).toBe('free')
			expect(shape.props.segments[2].type).toBe('straight')
		})

		it('Extends previously drawn line when shift is held', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(10, 10)
				.pointerUp()
				.pointerDown(20, 20)

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape1.props.segments.length).toBe(2)
			expect(shape1.props.segments[0].type).toBe('straight')
			expect(shape1.props.segments[1].type).toBe('straight')

			editor.pointerUp().pointerDown(30, 30).pointerUp()

			const shape2 = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape2.props.segments.length).toBe(3)
			expect(shape2.props.segments[2].type).toBe('straight')
		})

		it('Does not extends previously drawn line after switching to another tool', () => {
			editor
				.setCurrentTool(toolType)
				.pointerDown(10, 10)
				.pointerUp()
				.setCurrentTool('select')
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(20, 20)
				.pointerMove(30, 30)

			expect(editor.getCurrentPageShapes()).toHaveLength(2)

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape1.props.segments.length).toBe(1)
			expect(shape1.props.segments[0].type).toBe('free')

			const shape2 = editor.getCurrentPageShapes()[1] as DrawableShape
			expect(shape2.props.segments.length).toBe(1)
			expect(shape2.props.segments[0].type).toBe('straight')
		})

		it('Snaps to 15 degree angle when shift is held', () => {
			const magnitude = 10
			const angle = (17 * Math.PI) / 180
			const x = magnitude * Math.cos(angle)
			const y = magnitude * Math.sin(angle)

			const snappedAngle = (15 * Math.PI) / 180
			const snappedX = magnitude * Math.cos(snappedAngle)
			const snappedY = magnitude * Math.sin(snappedAngle)

			editor.setCurrentTool(toolType).keyDown('Shift').pointerDown(0, 0).pointerMove(x, y)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment = shape.props.segments[0]
			expect(segment.points[1].x).toBeCloseTo(snappedX)
			expect(segment.points[1].y).toBeCloseTo(snappedY)
		})

		it('Doesnt snap to 15 degree angle when cmd is held', () => {
			const magnitude = 10
			const angle = (17 * Math.PI) / 180
			const x = magnitude * Math.cos(angle)
			const y = magnitude * Math.sin(angle)

			editor.setCurrentTool(toolType).keyDown('Meta').pointerDown(0, 0).pointerMove(x, y)

			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment = shape.props.segments[0]
			expect(segment.points[1].x).toBeCloseTo(x)
			expect(segment.points[1].y).toBeCloseTo(y)
		})

		it('Snaps to start or end of straight segments in self when shift + cmd is held', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(0, 0)
				.pointerUp()
				.pointerDown(0, 10)
				.pointerUp()
				.pointerDown(10, 0)
				.pointerUp()
				.pointerDown(10, 0)
				.pointerMove(1, 0) // very close to first point

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment1 = last(shape1.props.segments)!
			const point1 = last(segment1.points)!
			expect(point1.x).toBe(1)

			editor.keyDown('Meta')
			const shape2 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment2 = last(shape2.props.segments)!
			const point2 = last(segment2.points)!
			expect(point2.x).toBe(0)
		})

		it('Snaps to position along straight segments in self when shift + cmd is held', () => {
			editor
				.setCurrentTool(toolType)
				.keyDown('Shift')
				.pointerDown(0, 0)
				.pointerUp()
				.pointerDown(0, 10)
				.pointerUp()
				.pointerDown(10, 5)
				.pointerUp()
				.pointerDown(10, 5)
				.pointerMove(1, 5)

			const shape1 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment1 = last(shape1.props.segments)!
			const point1 = last(segment1.points)!
			expect(point1.x).toBe(1)

			editor.keyDown('Meta')
			const shape2 = editor.getCurrentPageShapes()[0] as DrawableShape
			const segment2 = last(shape2.props.segments)!
			const point2 = last(segment2.points)!
			expect(point2.x).toBe(0)
		})

		it('Deletes very short lines on interrupt', () => {
			editor.setCurrentTool(toolType).pointerDown(0, 0).pointerMove(0.1, 0.1).interrupt()
			expect(editor.getCurrentPageShapes()).toHaveLength(0)
		})

		it('Does not delete longer lines on interrupt', () => {
			editor.setCurrentTool(toolType).pointerDown(0, 0).pointerMove(5, 5).interrupt()
			expect(editor.getCurrentPageShapes()).toHaveLength(1)
		})

		it('Completes on cancel', () => {
			editor.setCurrentTool(toolType).pointerDown(0, 0).pointerMove(5, 5).cancel()
			expect(editor.getCurrentPageShapes()).toHaveLength(1)
			const shape = editor.getCurrentPageShapes()[0] as DrawableShape
			expect(shape.props.segments.length).toBe(1)
		})
	})
}

it('Draws a bunch', () => {
	editor.setCurrentTool('draw').setCamera({ x: 0, y: 0, z: 1 })

	const [first, ...rest] = TEST_DRAW_SHAPE_SCREEN_POINTS
	editor.pointerMove(first.x, first.y).pointerDown()

	for (const point of rest) {
		editor.pointerMove(point.x, point.y)
	}

	editor.pointerUp()
	editor.selectAll()

	const shape = { ...editor.getLastCreatedShape() }
	// @ts-expect-error
	delete shape.id
	expect(shape).toMatchSnapshot('draw shape')
})
