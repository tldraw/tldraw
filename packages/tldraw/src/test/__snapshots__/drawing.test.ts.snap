// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Draws a bunch > draw shape 1`] = `
{
  "index": "a1",
  "isLocked": false,
  "meta": {},
  "opacity": 1,
  "parentId": "page:page",
  "props": {
    "color": "black",
    "dash": "draw",
    "fill": "none",
    "isClosed": false,
    "isComplete": true,
    "isPen": false,
    "scale": 1,
    "segments": [
      {
        "points": [
          {
            "x": 0,
            "y": 0,
            "z": 0.5,
          },
          {
            "x": 1,
            "y": 0,
            "z": 0.5,
          },
          {
            "x": 4,
            "y": 0,
            "z": 0.5,
          },
          {
            "x": 10,
            "y": -1,
            "z": 0.5,
          },
          {
            "x": 19,
            "y": -4,
            "z": 0.5,
          },
          {
            "x": 30,
            "y": -10,
            "z": 0.5,
          },
          {
            "x": 46,
            "y": -20,
            "z": 0.5,
          },
          {
            "x": 61,
            "y": -30,
            "z": 0.5,
          },
          {
            "x": 74,
            "y": -43,
            "z": 0.5,
          },
          {
            "x": 89,
            "y": -59,
            "z": 0.5,
          },
          {
            "x": 102,
            "y": -77,
            "z": 0.5,
          },
          {
            "x": 108,
            "y": -90,
            "z": 0.5,
          },
          {
            "x": 112,
            "y": -103,
            "z": 0.5,
          },
          {
            "x": 117,
            "y": -119,
            "z": 0.5,
          },
          {
            "x": 118,
            "y": -131,
            "z": 0.5,
          },
          {
            "x": 119,
            "y": -137,
            "z": 0.5,
          },
          {
            "x": 119,
            "y": -145,
            "z": 0.5,
          },
          {
            "x": 120,
            "y": -152,
            "z": 0.5,
          },
          {
            "x": 119,
            "y": -158,
            "z": 0.5,
          },
          {
            "x": 117,
            "y": -163,
            "z": 0.5,
          },
          {
            "x": 114,
            "y": -167,
            "z": 0.5,
          },
          {
            "x": 109,
            "y": -169,
            "z": 0.5,
          },
          {
            "x": 103,
            "y": -170,
            "z": 0.5,
          },
          {
            "x": 97,
            "y": -170,
            "z": 0.5,
          },
          {
            "x": 89,
            "y": -170,
            "z": 0.5,
          },
          {
            "x": 80,
            "y": -166,
            "z": 0.5,
          },
          {
            "x": 71,
            "y": -159,
            "z": 0.5,
          },
          {
            "x": 62,
            "y": -150,
            "z": 0.5,
          },
          {
            "x": 54,
            "y": -138,
            "z": 0.5,
          },
          {
            "x": 50,
            "y": -126,
            "z": 0.5,
          },
          {
            "x": 47,
            "y": -113,
            "z": 0.5,
          },
          {
            "x": 46,
            "y": -99,
            "z": 0.5,
          },
          {
            "x": 46,
            "y": -82,
            "z": 0.5,
          },
          {
            "x": 47,
            "y": -61,
            "z": 0.5,
          },
          {
            "x": 53,
            "y": -41,
            "z": 0.5,
          },
          {
            "x": 60,
            "y": -24,
            "z": 0.5,
          },
          {
            "x": 68,
            "y": -7,
            "z": 0.5,
          },
          {
            "x": 79,
            "y": 12,
            "z": 0.5,
          },
          {
            "x": 88,
            "y": 32,
            "z": 0.5,
          },
          {
            "x": 96,
            "y": 50,
            "z": 0.5,
          },
          {
            "x": 103,
            "y": 69,
            "z": 0.5,
          },
          {
            "x": 106,
            "y": 86,
            "z": 0.5,
          },
          {
            "x": 107,
            "y": 102,
            "z": 0.5,
          },
          {
            "x": 107,
            "y": 120,
            "z": 0.5,
          },
          {
            "x": 102,
            "y": 136,
            "z": 0.5,
          },
          {
            "x": 90,
            "y": 146,
            "z": 0.5,
          },
          {
            "x": 74,
            "y": 154,
            "z": 0.5,
          },
          {
            "x": 43,
            "y": 163,
            "z": 0.5,
          },
          {
            "x": 32,
            "y": 164,
            "z": 0.5,
          },
          {
            "x": 21,
            "y": 164,
            "z": 0.5,
          },
          {
            "x": 11,
            "y": 164,
            "z": 0.5,
          },
          {
            "x": 2,
            "y": 164,
            "z": 0.5,
          },
          {
            "x": -7,
            "y": 162,
            "z": 0.5,
          },
          {
            "x": -13,
            "y": 159,
            "z": 0.5,
          },
          {
            "x": -15,
            "y": 153,
            "z": 0.5,
          },
          {
            "x": -15,
            "y": 147,
            "z": 0.5,
          },
          {
            "x": -11,
            "y": 138,
            "z": 0.5,
          },
          {
            "x": 1,
            "y": 127,
            "z": 0.5,
          },
          {
            "x": 15,
            "y": 112,
            "z": 0.5,
          },
          {
            "x": 34,
            "y": 96,
            "z": 0.5,
          },
          {
            "x": 56,
            "y": 79,
            "z": 0.5,
          },
          {
            "x": 81,
            "y": 58,
            "z": 0.5,
          },
          {
            "x": 107,
            "y": 33,
            "z": 0.5,
          },
          {
            "x": 126,
            "y": 12,
            "z": 0.5,
          },
          {
            "x": 145,
            "y": -10,
            "z": 0.5,
          },
          {
            "x": 160,
            "y": -30,
            "z": 0.5,
          },
          {
            "x": 172,
            "y": -50,
            "z": 0.5,
          },
          {
            "x": 185,
            "y": -73,
            "z": 0.5,
          },
          {
            "x": 194,
            "y": -93,
            "z": 0.5,
          },
          {
            "x": 199,
            "y": -112,
            "z": 0.5,
          },
          {
            "x": 202,
            "y": -127,
            "z": 0.5,
          },
          {
            "x": 203,
            "y": -138,
            "z": 0.5,
          },
          {
            "x": 203,
            "y": -146,
            "z": 0.5,
          },
          {
            "x": 201,
            "y": -152,
            "z": 0.5,
          },
          {
            "x": 196,
            "y": -155,
            "z": 0.5,
          },
          {
            "x": 191,
            "y": -156,
            "z": 0.5,
          },
          {
            "x": 186,
            "y": -157,
            "z": 0.5,
          },
          {
            "x": 178,
            "y": -156,
            "z": 0.5,
          },
          {
            "x": 170,
            "y": -150,
            "z": 0.5,
          },
          {
            "x": 164,
            "y": -140,
            "z": 0.5,
          },
          {
            "x": 158,
            "y": -128,
            "z": 0.5,
          },
          {
            "x": 151,
            "y": -110,
            "z": 0.5,
          },
          {
            "x": 144,
            "y": -89,
            "z": 0.5,
          },
          {
            "x": 139,
            "y": -64,
            "z": 0.5,
          },
          {
            "x": 135,
            "y": -36,
            "z": 0.5,
          },
          {
            "x": 132,
            "y": -7,
            "z": 0.5,
          },
          {
            "x": 132,
            "y": 22,
            "z": 0.5,
          },
          {
            "x": 132,
            "y": 49,
            "z": 0.5,
          },
          {
            "x": 133,
            "y": 74,
            "z": 0.5,
          },
          {
            "x": 140,
            "y": 97,
            "z": 0.5,
          },
          {
            "x": 148,
            "y": 113,
            "z": 0.5,
          },
          {
            "x": 156,
            "y": 124,
            "z": 0.5,
          },
          {
            "x": 166,
            "y": 137,
            "z": 0.5,
          },
          {
            "x": 175,
            "y": 145,
            "z": 0.5,
          },
          {
            "x": 183,
            "y": 150,
            "z": 0.5,
          },
          {
            "x": 191,
            "y": 152,
            "z": 0.5,
          },
          {
            "x": 197,
            "y": 152,
            "z": 0.5,
          },
          {
            "x": 205,
            "y": 151,
            "z": 0.5,
          },
          {
            "x": 214,
            "y": 146,
            "z": 0.5,
          },
          {
            "x": 223,
            "y": 136,
            "z": 0.5,
          },
          {
            "x": 230,
            "y": 125,
            "z": 0.5,
          },
          {
            "x": 236,
            "y": 112,
            "z": 0.5,
          },
          {
            "x": 242,
            "y": 95,
            "z": 0.5,
          },
          {
            "x": 247,
            "y": 78,
            "z": 0.5,
          },
          {
            "x": 250,
            "y": 61,
            "z": 0.5,
          },
          {
            "x": 252,
            "y": 46,
            "z": 0.5,
          },
          {
            "x": 253,
            "y": 37,
            "z": 0.5,
          },
          {
            "x": 253,
            "y": 31,
            "z": 0.5,
          },
          {
            "x": 253,
            "y": 24,
            "z": 0.5,
          },
          {
            "x": 251,
            "y": 20,
            "z": 0.5,
          },
          {
            "x": 248,
            "y": 16,
            "z": 0.5,
          },
          {
            "x": 246,
            "y": 16,
            "z": 0.5,
          },
          {
            "x": 243,
            "y": 16,
            "z": 0.5,
          },
          {
            "x": 240,
            "y": 17,
            "z": 0.5,
          },
          {
            "x": 238,
            "y": 19,
            "z": 0.5,
          },
          {
            "x": 236,
            "y": 26,
            "z": 0.5,
          },
          {
            "x": 234,
            "y": 34,
            "z": 0.5,
          },
          {
            "x": 233,
            "y": 45,
            "z": 0.5,
          },
          {
            "x": 232,
            "y": 56,
            "z": 0.5,
          },
          {
            "x": 232,
            "y": 66,
            "z": 0.5,
          },
          {
            "x": 235,
            "y": 79,
            "z": 0.5,
          },
          {
            "x": 241,
            "y": 91,
            "z": 0.5,
          },
          {
            "x": 247,
            "y": 100,
            "z": 0.5,
          },
          {
            "x": 255,
            "y": 109,
            "z": 0.5,
          },
          {
            "x": 260,
            "y": 113,
            "z": 0.5,
          },
          {
            "x": 266,
            "y": 116,
            "z": 0.5,
          },
          {
            "x": 274,
            "y": 118,
            "z": 0.5,
          },
          {
            "x": 280,
            "y": 118,
            "z": 0.5,
          },
          {
            "x": 286,
            "y": 115,
            "z": 0.5,
          },
          {
            "x": 291,
            "y": 105,
            "z": 0.5,
          },
          {
            "x": 296,
            "y": 93,
            "z": 0.5,
          },
          {
            "x": 298,
            "y": 83,
            "z": 0.5,
          },
          {
            "x": 301,
            "y": 70,
            "z": 0.5,
          },
          {
            "x": 303,
            "y": 58,
            "z": 0.5,
          },
          {
            "x": 305,
            "y": 48,
            "z": 0.5,
          },
          {
            "x": 306,
            "y": 38,
            "z": 0.5,
          },
          {
            "x": 307,
            "y": 31,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 25,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 22,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 20,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 19,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 22,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 27,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 35,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 44,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 51,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 56,
            "z": 0.5,
          },
          {
            "x": 308,
            "y": 61,
            "z": 0.5,
          },
          {
            "x": 309,
            "y": 66,
            "z": 0.5,
          },
          {
            "x": 312,
            "y": 71,
            "z": 0.5,
          },
          {
            "x": 314,
            "y": 74,
            "z": 0.5,
          },
          {
            "x": 317,
            "y": 75,
            "z": 0.5,
          },
          {
            "x": 320,
            "y": 76,
            "z": 0.5,
          },
          {
            "x": 324,
            "y": 76,
            "z": 0.5,
          },
          {
            "x": 329,
            "y": 73,
            "z": 0.5,
          },
          {
            "x": 333,
            "y": 69,
            "z": 0.5,
          },
          {
            "x": 336,
            "y": 66,
            "z": 0.5,
          },
          {
            "x": 339,
            "y": 62,
            "z": 0.5,
          },
          {
            "x": 342,
            "y": 59,
            "z": 0.5,
          },
          {
            "x": 344,
            "y": 57,
            "z": 0.5,
          },
          {
            "x": 346,
            "y": 55,
            "z": 0.5,
          },
          {
            "x": 348,
            "y": 55,
            "z": 0.5,
          },
          {
            "x": 348,
            "y": 55,
            "z": 0.5,
          },
          {
            "x": 349,
            "y": 55,
            "z": 0.5,
          },
          {
            "x": 349,
            "y": 56,
            "z": 0.5,
          },
          {
            "x": 350,
            "y": 57,
            "z": 0.5,
          },
          {
            "x": 351,
            "y": 59,
            "z": 0.5,
          },
          {
            "x": 351,
            "y": 61,
            "z": 0.5,
          },
          {
            "x": 352,
            "y": 62,
            "z": 0.5,
          },
          {
            "x": 352,
            "y": 63,
            "z": 0.5,
          },
          {
            "x": 353,
            "y": 64,
            "z": 0.5,
          },
          {
            "x": 354,
            "y": 64,
            "z": 0.5,
          },
          {
            "x": 355,
            "y": 64,
            "z": 0.5,
          },
          {
            "x": 356,
            "y": 58,
            "z": 0.5,
          },
          {
            "x": 358,
            "y": 49,
            "z": 0.5,
          },
          {
            "x": 360,
            "y": 40,
            "z": 0.5,
          },
          {
            "x": 363,
            "y": 32,
            "z": 0.5,
          },
          {
            "x": 365,
            "y": 26,
            "z": 0.5,
          },
          {
            "x": 367,
            "y": 19,
            "z": 0.5,
          },
          {
            "x": 369,
            "y": 13,
            "z": 0.5,
          },
          {
            "x": 373,
            "y": 7,
            "z": 0.5,
          },
          {
            "x": 376,
            "y": 3,
            "z": 0.5,
          },
          {
            "x": 380,
            "y": 2,
            "z": 0.5,
          },
          {
            "x": 385,
            "y": 2,
            "z": 0.5,
          },
          {
            "x": 390,
            "y": 2,
            "z": 0.5,
          },
          {
            "x": 397,
            "y": 3,
            "z": 0.5,
          },
          {
            "x": 410,
            "y": 11,
            "z": 0.5,
          },
          {
            "x": 424,
            "y": 23,
            "z": 0.5,
          },
          {
            "x": 434,
            "y": 34,
            "z": 0.5,
          },
          {
            "x": 446,
            "y": 49,
            "z": 0.5,
          },
          {
            "x": 456,
            "y": 64,
            "z": 0.5,
          },
          {
            "x": 464,
            "y": 81,
            "z": 0.5,
          },
          {
            "x": 468,
            "y": 95,
            "z": 0.5,
          },
          {
            "x": 470,
            "y": 116,
            "z": 0.5,
          },
          {
            "x": 472,
            "y": 142,
            "z": 0.5,
          },
          {
            "x": 472,
            "y": 162,
            "z": 0.5,
          },
          {
            "x": 468,
            "y": 178,
            "z": 0.5,
          },
          {
            "x": 458,
            "y": 195,
            "z": 0.5,
          },
          {
            "x": 442,
            "y": 213,
            "z": 0.5,
          },
          {
            "x": 423,
            "y": 230,
            "z": 0.5,
          },
          {
            "x": 407,
            "y": 240,
            "z": 0.5,
          },
          {
            "x": 393,
            "y": 245,
            "z": 0.5,
          },
          {
            "x": 377,
            "y": 250,
            "z": 0.5,
          },
          {
            "x": 364,
            "y": 252,
            "z": 0.5,
          },
          {
            "x": 354,
            "y": 252,
            "z": 0.5,
          },
          {
            "x": 346,
            "y": 248,
            "z": 0.5,
          },
          {
            "x": 340,
            "y": 239,
            "z": 0.5,
          },
          {
            "x": 339,
            "y": 225,
            "z": 0.5,
          },
          {
            "x": 339,
            "y": 198,
            "z": 0.5,
          },
          {
            "x": 349,
            "y": 165,
            "z": 0.5,
          },
          {
            "x": 372,
            "y": 130,
            "z": 0.5,
          },
          {
            "x": 403,
            "y": 89,
            "z": 0.5,
          },
          {
            "x": 432,
            "y": 54,
            "z": 0.5,
          },
          {
            "x": 467,
            "y": 16,
            "z": 0.5,
          },
          {
            "x": 504,
            "y": -21,
            "z": 0.5,
          },
          {
            "x": 551,
            "y": -68,
            "z": 0.5,
          },
          {
            "x": 597,
            "y": -115,
            "z": 0.5,
          },
          {
            "x": 619,
            "y": -138,
            "z": 0.5,
          },
          {
            "x": 641,
            "y": -162,
            "z": 0.5,
          },
          {
            "x": 663,
            "y": -188,
            "z": 0.5,
          },
          {
            "x": 675,
            "y": -203,
            "z": 0.5,
          },
          {
            "x": 684,
            "y": -219,
            "z": 0.5,
          },
          {
            "x": 692,
            "y": -237,
            "z": 0.5,
          },
          {
            "x": 693,
            "y": -244,
            "z": 0.5,
          },
          {
            "x": 691,
            "y": -250,
            "z": 0.5,
          },
          {
            "x": 682,
            "y": -254,
            "z": 0.5,
          },
          {
            "x": 664,
            "y": -256,
            "z": 0.5,
          },
          {
            "x": 642,
            "y": -256,
            "z": 0.5,
          },
          {
            "x": 621,
            "y": -253,
            "z": 0.5,
          },
          {
            "x": 589,
            "y": -240,
            "z": 0.5,
          },
          {
            "x": 554,
            "y": -221,
            "z": 0.5,
          },
          {
            "x": 526,
            "y": -201,
            "z": 0.5,
          },
          {
            "x": 502,
            "y": -182,
            "z": 0.5,
          },
          {
            "x": 484,
            "y": -165,
            "z": 0.5,
          },
          {
            "x": 467,
            "y": -146,
            "z": 0.5,
          },
          {
            "x": 456,
            "y": -131,
            "z": 0.5,
          },
          {
            "x": 450,
            "y": -120,
            "z": 0.5,
          },
          {
            "x": 448,
            "y": -112,
            "z": 0.5,
          },
          {
            "x": 448,
            "y": -107,
            "z": 0.5,
          },
          {
            "x": 449,
            "y": -104,
            "z": 0.5,
          },
          {
            "x": 452,
            "y": -103,
            "z": 0.5,
          },
          {
            "x": 458,
            "y": -102,
            "z": 0.5,
          },
          {
            "x": 462,
            "y": -102,
            "z": 0.5,
          },
          {
            "x": 465,
            "y": -103,
            "z": 0.5,
          },
          {
            "x": 470,
            "y": -104,
            "z": 0.5,
          },
          {
            "x": 472,
            "y": -105,
            "z": 0.5,
          },
          {
            "x": 474,
            "y": -106,
            "z": 0.5,
          },
          {
            "x": 475,
            "y": -106,
            "z": 0.5,
          },
          {
            "x": 476,
            "y": -107,
            "z": 0.5,
          },
          {
            "x": 476,
            "y": -107,
            "z": 0.5,
          },
          {
            "x": 477,
            "y": -107,
            "z": 0.5,
          },
        ],
        "type": "free",
      },
    ],
    "size": "m",
  },
  "rotation": 0,
  "type": "draw",
  "typeName": "shape",
  "x": 511,
  "y": 234,
}
`;
